File: backend/internal/handlers/handlers.go
package handlers

import (
	"net/http"

	"github.com/Gunter-Q12/lns/backend/internal/executor"
	"github.com/gin-gonic/gin"
	"go.uber.org/zap"
)

func Map(r *gin.Engine, h *Handlers) {
	r.GET("/api/*path", h.Get)
}

type Handlers struct {
	logger *zap.Logger
	exe    *executor.Executor
}

func New(exe *executor.Executor, logger *zap.Logger) *Handlers {
	return &Handlers{exe: exe, logger: logger}
}

func (h *Handlers) Get(c *gin.Context) {
	var nsCmd []string
	var cmd []string

	namespace := c.Query("namespace")
	if namespace != "" {
		nsCmd = []string{"netns", "--net=" + namespace}
	}

	switch c.Param("path") {
	case "/namespaces":
		cmd = []string{"lsns", "--json", "-t", "net"}
	case "/nft":
		cmd = []string{"nft", "--json", "list", "ruleset"}
	case "/route":
		cmd = []string{"ip", "--json", "route", "show", "table", "all"}
	case "/rule4":
		cmd = []string{"ip", "--json", "rule"}
	case "/rule6":
		cmd = []string{"ip", "-6", "--json", "rule"}
	case "/addr":
		cmd = []string{"ip", "--json", "addr"}
	default:
		c.String(http.StatusNotFound, "Received unknown path: %s", c.Request.URL.Path)
		return
	}

	finalCmd := append(nsCmd, cmd...)

	info, err := h.exe.Run(c.Request.Context(), finalCmd[0], finalCmd[1:]...)
	if err != nil {
		h.logger.Error("getting data", zap.Error(err))
		c.JSON(http.StatusInternalServerError, gin.H{"errors": err.Error()})
		return
	}

	c.Data(http.StatusOK, "application/json", info)
}

File: backend-py/app/internal/handlers/handlers.py
from fastapi import APIRouter, Request, HTTPException, Query, Path, Response
from app.internal.executor.executor import Executor
import structlog
import json

router = APIRouter()
logger = structlog.get_logger()
executor = Executor()

@router.get("/api/{path:path}")
async def get_path(
    path: str,
    namespace: str = Query(None)
):
    ns_cmd = []
    if namespace:
        ns_cmd = ["nsenter", f"--net={namespace}"]

    cmd = []
    if path == "namespaces" or path == "/namespaces":
        cmd = ["lsns", "--json", "-t", "net"]
    elif path == "nft" or path == "/nft":
        cmd = ["nft", "--json", "list", "ruleset"]
    elif path == "route" or path == "/route":
        cmd = ["ip", "--json", "route", "show", "table", "all"]
    elif path == "rule4" or path == "/rule4":
        cmd = ["ip", "--json", "rule"]
    elif path == "rule6" or path == "/rule6":
        cmd = ["ip", "-6", "--json", "rule"]
    elif path == "addr" or path == "/addr":
        cmd = ["ip", "--json", "addr"]
    else:
        raise HTTPException(status_code=404, detail=f"Received unknown path: /api/{path}")

    final_cmd = ns_cmd + cmd

    try:
        output = await executor.run(final_cmd[0], *final_cmd[1:])

        try:
            return Response(content=output, media_type="application/json")
        except Exception as e:
            logger.error("failed to process command output", error=str(e))
            raise HTTPException(status_code=500, detail="Internal server error processing command output")

    except Exception as e:
        logger.error("getting data", error=str(e))
        raise HTTPException(status_code=500, detail=str(e))
